Mathlib/Topology/Algebra/ContinuousMonoidHom.lean

```diff
@@ -376,4 +379,75 @@ def compRight {B : Type*} [CommGroup B] [TopologicalSpace B] [TopologicalGroup B
 #align continuous_monoid_hom.comp_right ContinuousMonoidHom.compRight
 #align continuous_add_monoid_hom.comp_right ContinuousAddMonoidHom.compRight
 
+section LocallyCompact
+
+variable {X Y : Type*} [TopologicalSpace X] [Group X] [TopologicalGroup X]
+  [UniformSpace Y] [CommGroup Y] [UniformGroup Y] [T0Space Y] [CompactSpace Y]
+
+theorem locallyCompactSpace_of_equicontinuousAt (U : Set X) (V : Set Y)
+    (hU : IsCompact U) (hV : V ∈ nhds (1 : Y))
+    (h : EquicontinuousAt (fun f : {f : X →* Y | Set.MapsTo f U V} ↦ (f : X → Y)) 1) :
+    LocallyCompactSpace (ContinuousMonoidHom X Y) := by
+  replace h := equicontinuous_of_equicontinuousAt_one _ h
+  obtain ⟨W, hWo, hWV, hWc⟩ := local_compact_nhds hV
+  let S1 : Set (X →* Y) := {f | Set.MapsTo f U W}
+  let S2 : Set (ContinuousMonoidHom X Y) := {f | Set.MapsTo f U W}
+  let S3 : Set C(X, Y) := (↑) '' S2
+  let S4 : Set (X → Y) := (↑) '' S3
+  replace h : Equicontinuous ((↑) : S1 → X → Y) :=
+    h.comp (Subtype.map _root_.id fun f hf ↦ hf.mono_right hWV)
+  have hS4 : S4 = (↑) '' S1 := by
+    ext
+    constructor
+    · rintro ⟨-, ⟨f, hf, rfl⟩, rfl⟩
+      exact ⟨f, hf, rfl⟩
+    · rintro ⟨f, hf, rfl⟩
+      exact ⟨⟨f, h.continuous ⟨f, hf⟩⟩, ⟨⟨f, h.continuous ⟨f, hf⟩⟩, hf, rfl⟩, rfl⟩
+  replace h : Equicontinuous ((↑) : S3 → X → Y) := by
+    rw [equicontinuous_iff_range, ← Set.image_eq_range] at h ⊢
+    rwa [← hS4] at h
+  replace hS4 : S4 = Set.pi U (fun _ ↦ W) ∩ Set.range ((↑) : (X →* Y) → (X → Y)) := by
+    simp_rw [hS4, Set.ext_iff, Set.mem_image, S1, Set.mem_setOf_eq]
+    exact fun f ↦ ⟨fun ⟨g, hg, hf⟩ ↦ hf ▸ ⟨hg, g, rfl⟩, fun ⟨hg, g, hf⟩ ↦ ⟨g, hf ▸ hg, hf⟩⟩
+  replace hS4 : IsClosed S4 :=
+    hS4.symm ▸ (isClosed_set_pi (fun _ _ ↦ hWc.isClosed)).inter (MonoidHom.isClosed_range_coe X Y)
+  have hS2 : (interior S2).Nonempty := by
+    let T : Set (ContinuousMonoidHom X Y) := {f | Set.MapsTo f U (interior W)}
+    have h1 : T.Nonempty := ⟨1, fun _ _ ↦ mem_interior_iff_mem_nhds.mpr hWo⟩
+    have h2 : T ⊆ S2 := fun f hf ↦ hf.mono_right interior_subset
+    have h3 : IsOpen T := isOpen_induced (ContinuousMap.isOpen_setOf_mapsTo hU isOpen_interior)
+    exact h1.mono (interior_maximal h2 h3)
+  exact TopologicalSpace.PositiveCompacts.locallyCompactSpace_of_group
+    ⟨⟨S2, (inducing_toContinuousMap X Y).isCompact_iff.mpr
+      (ArzelaAscoli.isCompact_of_equicontinuous S3 hS4.isCompact h)⟩, hS2⟩
+
+variable [LocallyCompactSpace X]
+
+theorem locallyCompactSpace_of_hasBasis (V : ℕ → Set Y)
+    (hV : ∀ {n x}, x ∈ V n → x * x ∈ V n → x ∈ V (n + 1))
```

Comment:
How related is this to [TopologicalGroup.exists_antitone_basis_nhds_one](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Topology/Algebra/Group/Basic.html#TopologicalGroup.exists_antitone_basis_nhds_one)? I'm just wondering if it's possible to get away with a more natural condition here at all like `FirstCountableTopology`. The answer is likely no, but I thought I'd ask.

Reply:
Yeah, I think I thought about this possibility, but the condition in docs#TopologicalGroup.exists_antitone_basis_nhds_one isn't enough to get the condition in `hV` (For instance, think about the circle where the first neighborhood is the whole circle and the second neighborhood is the right half of the circle. This satisfies the condition in docs#TopologicalGroup.exists_antitone_basis_nhds_one but not the condition in `hV`). And I think you need a condition like `hV` in order to prove equicontinuity.

Reply:
Can you prove `hV` from first countability? It just feels like these are somehow very closely connected to me.

Reply:
Hmm, I had thought not, but now I'm not so sure. I guess if you take V(n+1) = Vn cap sq^(-1) (Vn), then this has the property that x in Vn and x^2 in Vn implies x in V(n+1). And you can probably further intersect with the (n+1)st term of an existing countable basis to finish the job?

Reply:
Ah, the issue is that if you shrink V(n+1) then hV no longer holds. And I wasn't able to find a way around this.

Reply:
Without some more clever way of modifying the sequence, you would need to find a V0 such that iterating V(n+1) = Vn cap sq^(-1) (Vn) gives a basis.

Reply:
Fair enough. That makes sense. It seemed too good to expect. I didn't see a way to do it either.